apps/src/blockly/addons/cdoBlockFlyout.js

```diff
@@ -0,0 +1,212 @@
+   * @returns The flyout's SVG group.
+   */
+  createDom(tagName) {
+    super.createDom(tagName);
+    tagName =
+      'flyoutClip' +
+      Blockly.utils.idGenerator.genUid().replace(/([\(\)])/g, '');
+    let definitions = Blockly.utils.dom.createSvgElement(
+      'defs',
+      {},
+      this.svgGroup_
+    );
+    definitions = Blockly.utils.dom.createSvgElement(
+      'clipPath',
+      {
+        id: tagName,
+      },
+      definitions
+    );
+    this.svgClipPath_ = Blockly.utils.dom.createSvgElement(
+      'path',
+      {},
+      definitions
+    );
+    this.svgGroup_.setAttribute('clip-path', 'url(#' + tagName + ')');
+    this.svgGroup_.classList.add('blockFieldFlyout');
+    return this.svgGroup_;
+  }
+
+  /**
+   * Initializes the flyout.
+   *
+   * @param targetWorkspace The workspace in which to
+   *     create new blocks.
+   */
+  init(targetWorkspace) {
+    super.init(targetWorkspace);
+    this.targetWorkspace_ = targetWorkspace;
+  }
+
+  /**
+   * Compute height of flyout.  Position mat under each block.
+   * For RTL: Lay out the blocks right-aligned.
+   */
+  reflowInternal_() {
+    this.height_ = 0;
+    this.width_ = 0;
+    const topBlocks = this.workspace_.getTopBlocks(false);
+    for (let i = 0, block; (block = topBlocks[i]); i++) {
+      const blockHW = block.getHeightWidth();
+      this.updateHeight_(blockHW.height);
+      this.updateWidth_(blockHW.width);
+      if (this.rectMap_.has(block)) {
+        this.moveRectToBlock_(this.rectMap_.get(block), block);
+      }
+    }
+    // We need to set the flyout width based on the block itself, excluding children.
+    // Using block.getHeightWidth() would include blocks connected to input and next
+    // connections, which could make the flyout wider than the block that contains it.
+    const topBlockWidth = this.sourceBlock_.svgGroup_
+      .querySelector('path')
+      .getBoundingClientRect().width;
+    this.width_ = Math.max(this.width_, topBlockWidth - 36);
+    this.setBackgroundPath_(this.width_, this.height_);
+    this.position();
+  }
+
+  /** Update the flyout height based on the new block height.
+   *
+   * @param {number} newHeight - The new block height.
+   * @private
+   */
+  updateHeight_(newHeight) {
+    this.height_ = Math.max(this.height_, newHeight);
+  }
+  /**
+   * Update the flyout width based on the new block width.
+   *
+   * @param {number} newWidth - The new block width.
+   * @private
+   */
+  updateWidth_(newWidth) {
+    this.width_ += newWidth + this.GAP_X;
+  }
+
+  /**
+   * Position the flyout.
+   */
+  position() {
+    this.isVisible() && this.positionAt_(this.width_, this.height_, 0, 0);
+  }
+
+  /**
+   * Create and set the path for the visible boundaries of the flyout.
+   *
+   * @param width The width of the flyout, not including the rounded corners.
+   * @param height The height of the flyout, not including rounded corners.
+   */
+  setBackgroundPath_(width, height) {
```

Comment:
Is this function replacing `setBackgroundPath` from `flyout_horizontal.ts` or is it an additional function needed for our customized flyout field?

Reply:
It's an override replacing the original `setBackgroundPath_`. I had to modify the override to make it work for our blocks when in RTL mode and also implemented Blockly's new SVG util commands.

Reply:
Meta question: what is the significance of functions with a `_` at the end versus not? Are functions with `_` overrides?

Reply:
Mike and I chatted about this offline - the change of name from `setBackgroundPath_` to `setBackgroundPath` happened [very recently](https://github.com/google/blockly/commit/48bdeb3bdb0a222a6b48589310343248bb29bcb1) in mainline Blockly (between v9 and v10). `code-dot-org` currently uses v9 but once we bump to v10, the function name will need to be updated (remove underscore). I was referring to v10 [here](https://github.com/google/blockly/blob/6b3803e68fd1e57ab20e0aaabe1be8dfc8177d24/core/flyout_horizontal.ts#L152) while reviewing which is why I asked the question above.

Reply:
@molly-moen here's the docs Maribeth directed me to when I had the same question:

https://developers.google.com/blockly/guides/configure/advanced/using_blockly_apis#visibility

Here's a note on underscores there:

> Note: Previously, Blockly used an underscore to denote private properties: example_. In keeping with the current Google style guide, we no longer do this in new code, and may change old code. Before calling a method in Blockly, ensure that it appears in the reference documentation or is not marked private in the code. In rare cases private functions were used so widely that they became public, so you may see public functions that end in an underscore.

Reply:
To add to what Ben shared, you can generally assume that functions ending in `_` are overrides, since that's not an internal convention we're trying to set for our own reasons. I've been trying to learn about and incorporate JSDocs as I work on these classes. If it would be helpful and appropriate, I could add an `@override` property to each function that also shows up in the parent class. Let me know your thoughts!

Reply:
@fisher-alice - we chatted offline, but i'm going to add these tags to the next branch that includes these changes. in progress now!